### PR TITLE
CLI bug fixes and enhancements

### PR DESCRIPTION
**Pull request by KPrasch**
##### Fixes Issues
- #614 - Question: Do we want this condition to `raise`?
- #634 - Unskip `nucypher ursula run` CLI test (Reactor / Pytest Issue)
- #610 - Timestamp temp system filepaths
- #551 - Auto-generate configuration root directory if it does not exist.
- #578 (Half) - Disable logging via CLI (`--quiet` / `-Q`)
- #583 - Allow destruction of invalid system configurations

Also attempts to address #582 

 ##### Shorthand CLI usage:

| Option | Shorthand |
-----------|----------------|
| `--dev` | `-d` |
| `--federated-only` | `-F` |
| `--debug` | `-D` |
| `--quiet` | `-Q` |
| `--force` | - |
| `--dry-run` | `-x` |

##### Other 
- Remove `cleanup` pytest fixture

### CODE REVIEW

File: nucypher/cli/main.py [hunk 1]
```diff
@@ -140,15 +143,16 @@ def status(click_config, config_file):
 @click.argument('action')
 @click.option('--debug', '-D', help="Enable debugging mode", is_flag=True)
 @click.option('--dev', '-d', help="Enable development mode", is_flag=True)
-@click.option('--dry-run', help="Execute normally without actually starting the node", is_flag=True)
+@click.option('--quiet', '-Q', help="Disable logging", is_flag=True)
```
Comment (tuxxy) [inline nucypher/cli/main.py]: heh

File: nucypher/cli/main.py [hunk 2]
```diff
@@ -140,15 +143,16 @@ def status(click_config, config_file):
...
+@click.option('--dry-run', '-x', help="Execute normally without actually starting the node", is_flag=True)
 @click.option('--force', '-f', help="Don't ask for confirmation", is_flag=True)
 @click.option('--teacher-uri', help="An Ursula URI to start learning from (seednode)", type=click.STRING)
 @click.option('--min-stake', help="The minimum stake the teacher must have to be a teacher", type=click.INT, default=0)
 @click.option('--rest-host', help="The host IP address to run Ursula network services on", type=click.STRING)
 @click.option('--rest-port', help="The host port to run Ursula network services on", type=NETWORK_PORT)
 @click.option('--db-filepath', help="The database filepath to connect to", type=click.STRING)
 @click.option('--checksum-address', help="Run with a specified account", type=EIP55_CHECKSUM_ADDRESS)
-@click.option('--federated-only', help="Connect only to federated nodes", is_flag=True, default=FEDERATED_ONLY)
+@click.option('--federated-only', '-F', help="Connect only to federated nodes", is_flag=True, default=FEDERATED_ONLY)
```
Comment (tuxxy) [inline nucypher/cli/main.py]: I'd be careful with shadowing `-f`. Seems kinda spooky.

File: nucypher/cli/main.py [hunk 3]
```diff
@@ -197,31 +202,38 @@ def ursula(click_config,
...
 
-    if debug:
+    if debug and not quiet:
         click_config.log_to_sentry = False
         click_config.log_to_file = True
         globalLogPublisher.removeObserver(logToSentry)                          # Sentry
         globalLogPublisher.addObserver(SimpleObserver(log_level_name='debug'))  # Print
 
+    elif quiet:
+        globalLogPublisher.removeObserver(logToSentry)
```
Comment (tuxxy) [inline nucypher/cli/main.py]: This is neat.

File: nucypher/cli/main.py [hunk 4]
```diff
@@ -242,17 +254,71 @@ def ursula(click_config,
...
 
-        # Give the use a suggestion as to what to do next...
-        how_to_run_message = "\nTo run an Ursula node from the default configuration filepath run: \n\n'{}'\n"
-        suggested_command = 'nucypher ursula run'
-        if config_root is not None:
-            config_file_location = os.path.join(config_root, config_file or UrsulaConfiguration.CONFIG_FILENAME)
-            suggested_command += ' --config-file {}'.format(config_file_location)
-        click.secho(how_to_run_message.format(suggested_command), fg='green')
-        return  # FIN
+        if dev:
```
Comment (tuxxy) [inline nucypher/cli/main.py]: With all these conditionals, maybe it's best to start pulling out some of this logic into functions that are called to reduce lines. What do you think?

File: nucypher/cli/main.py [hunk 5]
```diff
@@ -242,17 +254,71 @@ def ursula(click_config,
...
+                message = 'Failed: No nucypher files found at {}'.format(ursula_config.config_root)
+                click.secho(message, fg='red')
+                log.debug(message)
+                raise click.Abort()
+            else:
+                message = "Deleted configuration files at {}".format(ursula_config.config_root)
+                click.secho(message, fg='green')
+                log.debug(message)
+
+        if not quiet:
```
Comment (tuxxy) [inline nucypher/cli/main.py]: Is there not a better way to echo without checking if we're `quiet` each and every time?

File: nucypher/cli/main.py (hunk 4 above)
Comment (KPrasch) [reply nucypher/cli/main.py]: This code has been subject to that cycle sever times now, not that it cant be improved more; However, I actually think `cli.main` is pretty concise and readable.  

No less, Im curious what your suggestions might be.

File: nucypher/cli/main.py (hunk 5 above)
Comment (KPrasch) [reply nucypher/cli/main.py]: @tuxxy - In this particular case case - I see what you mean - we can wrap `click.secho` and apply some sort of filtration, for example.

File: nucypher/cli/main.py (hunk 3 above)
Comment (KPrasch) [reply nucypher/cli/main.py]: Hopefully it's enough - I actually on worried about the possibility of lingering observers, so there is *no* guarantee that logging is disabled at the moment.

File: nucypher/cli/main.py (hunk 2 above)
Comment (KPrasch) [reply nucypher/cli/main.py]: Yeah - Perhaps let's not have `-f` and require the long form `--force`?

File: nucypher/cli/main.py (hunk 1 above)
Comment (KPrasch) [reply nucypher/cli/main.py]: I think this is too terse. How about: "Disable terminal echoing and logging"  or "We'll do it live" :joy:

File: nucypher/cli/main.py (hunk 2 above)
Comment (tuxxy) [reply nucypher/cli/main.py]: Yeah, that makes sense. Any objections from anyone else?

File: nucypher/cli/main.py (hunk 1 above)
Comment (tuxxy) [reply nucypher/cli/main.py]: Haha, the former is nice. :p

File: nucypher/cli/main.py (hunk 2 above)
Comment (cygnusv) [reply nucypher/cli/main.py]: +1

File: nucypher/cli/main.py [hunk 6]
```diff
@@ -198,31 +200,38 @@ def ursula(click_config,
     #
     # Boring Setup Stuff
     #
-    log = Logger('ursula.cli')
+    if not quiet:
+        log = Logger('ursula.cli')
 
-    if debug:
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/main.py]: Hmm - Yeah I like this idea.

File: nucypher/cli/main.py [hunk 7]
```diff
@@ -340,28 +354,31 @@ def ursula(click_config,
...
         finally:
-            click.secho("Stopping Ursula")
+            if not quiet:
+                click.secho("Stopping Ursula")
             ursula_config.cleanup()
-            click.secho("Ursula Stopped", fg='red')
+            if not quiet:
+                click.secho("Ursula Stopped", fg='red')
+            else:
+                click.secho("Stopped")
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/main.py]: heh..... I .... yeah.

File: nucypher/cli/main.py (hunk 5 above)
Comment (michwill) [reply nucypher/cli/main.py]: Yes, for example having a callback for `--quiet` which sets an option for this wrapper to not `secho` if it is `True`

File: nucypher/cli/main.py (hunk 2 above)
Comment (tuxxy) [reply nucypher/cli/main.py]: @KPrasch any changes in thought on this?

File: nucypher/cli/main.py [hunk 8]
```diff
@@ -140,15 +143,16 @@ def status(click_config, config_file):
 @click.argument('action')
 @click.option('--debug', '-D', help="Enable debugging mode", is_flag=True)
 @click.option('--dev', '-d', help="Enable development mode", is_flag=True)
-@click.option('--dry-run', help="Execute normally without actually starting the node", is_flag=True)
+@click.option('--quiet', '-Q', help="Disable logging", is_flag=True)
+@click.option('--dry-run', '-x', help="Execute normally without actually starting the node", is_flag=True)
 @click.option('--force', '-f', help="Don't ask for confirmation", is_flag=True)
```
Comment (tuxxy) [inline nucypher/cli/main.py]:
```suggestion
@click.option('--force', help="Don't ask for confirmation", is_flag=True)```

File: nucypher/cli/main.py (hunk 5 above)
Comment (jMyles) [reply nucypher/cli/main.py]: Or even patching `secho` depending on the flag.

But yeah, one way or another, the quiet checks and resulting indentation make it much harder to quickly read and review this part.

File: nucypher/config/node.py [hunk 9]
```diff
@@ -396,7 +397,9 @@ def validate(self, config_root: str, no_registry=False) -> bool:
 
         for field, path in filepaths.items():
             if not os.path.exists(path):
-                message = 'Missing configuration directory {}.'
+                message = 'Missing configuration file or directory: {}.'
+                if 'registry' in path:
```
Comment (jMyles) [inline nucypher/config/node.py]: Maybe a comment here describing to the reader *why* we expect 'registry' if they perhaps meant `--federated-only`?

File: nucypher/cli/main.py (hunk 2 above)
Comment (KPrasch) [reply nucypher/cli/main.py]: :heavy_check_mark:

File: nucypher/cli/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/main.py]: :heavy_check_mark:

File: nucypher/cli/main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/main.py]: :heavy_check_mark:

File: nucypher/config/node.py (hunk 9 above)
Comment (KPrasch) [reply nucypher/config/node.py]: :heavy_check_mark:

**Pull request merged**